chain/ton/client.go

```diff
@@ -0,0 +1,512 @@
+	if err != nil {
+		return "", fmt.Errorf("could not resolve token master address: %v", err)
+	}
+	if len(resp.JettonTransfers) == 0 {
+		return "", fmt.Errorf("could not resolve token master address: no transfer history")
+	}
+	masterAddr, err := tonaddress.ParseAddress(xc.Address(resp.JettonTransfers[0].JettonMaster))
+	if err != nil {
+		return "", err
+	}
+	return xc.ContractAddress(masterAddr.String()), nil
+}
+
+// This detects any JettonMessage in the nest of "InternalMessage"
+// This may need to be expanded as Jetton transfer could be nested deeper in more 'InternalMessages'
+func (client *Client) DetectJettonMovements(tx *api.Transaction) ([]*xc.LegacyTxInfoEndpoint, []*xc.LegacyTxInfoEndpoint, error) {
+	boc, err := base64.StdEncoding.DecodeString(tx.InMsg.MessageContent.Body)
+	if err != nil {
+		return nil, nil, err
+	}
+	inMsg, err := cell.FromBOC(boc)
+	if err != nil {
+		return nil, nil, err
+	}
+	inMsgMsg, err := inMsg.BeginParse().LoadRefCell()
+	if err != nil {
+		return nil, nil, err
+	}
+	internalMsg := &tlb.InternalMessage{}
+	err = tlb.LoadFromCell(internalMsg, inMsgMsg.BeginParse())
+	if err != nil {
+		return nil, nil, err
+	}
+	if internalMsg.Body == nil {
+		logrus.Debug("no jetton transfer detected (no body)")
+		return nil, nil, nil
+	}
+
+	jettonTfMaybe := &jetton.TransferPayload{}
+	err = tlb.LoadFromCell(jettonTfMaybe, internalMsg.Body.BeginParse())
+	if err != nil {
+		// give up here - no jetton movement(s)
+		logrus.WithError(err).Debug("no jetton transfer detected")
+		return nil, nil, nil
+	}
+	memo, ok := ParseComment(jettonTfMaybe.ForwardPayload)
+	fmt.Println("memo ", memo, ok)
+	if !ok {
+		memo, _ = ParseComment(jettonTfMaybe.CustomPayload)
+	}
+	tokenWallet := internalMsg.DstAddr
+	tokenMasterAddr, err := client.LookupJettonMasterForTokenWallet(tokenWallet)
+	if err != nil {
+		return nil, nil, err
+	}
+	chain := client.Asset.GetChain().Chain
+	amount := xc.AmountBlockchain(*jettonTfMaybe.Amount.Nano())
+	sources := []*xc.LegacyTxInfoEndpoint{
+		{
+			// this is the token wallet of the sender/owner
+			Address:         xc.Address(tokenWallet.String()),
+			Amount:          amount,
+			ContractAddress: tokenMasterAddr,
+			NativeAsset:     chain,
+			Memo:            memo,
+		},
+	}
+
+	dests := []*xc.LegacyTxInfoEndpoint{
+		{
+			Address:         xc.Address(jettonTfMaybe.Destination.String()),
+			Amount:          amount,
+			ContractAddress: tokenMasterAddr,
+			NativeAsset:     chain,
+			Memo:            memo,
+		},
+	}
+
+	return sources, dests, nil
+}
+
+// Prioritize getting tx by msg-hash as it's deterministic offline.  Fallback to using chain-calculated tx hash.
+func (client *Client) FetchTonTxByHash(ctx context.Context, txHash xc.TxHash) (api.Transaction, error) {
+	transactions := &api.TransactionsData{}
+
+	err := client.get(fmt.Sprintf("api/v3/transactionsByMessage?direction=in&msg_hash=%s", url.QueryEscape(string(txHash))), transactions)
```

Comment:
Why do we limit it by `direction=in` here?

Reply:
Adding comment.  TON has concept of output transactions where the chain makes outputs in response to input transactions.  The "inputs" are the ones clients actually submitted.